compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/CallGenerator.kt

```diff
@@ -328,7 +331,18 @@ internal class CallGenerator(
             is IrFieldExpressionBase -> {
                 val ownerField = symbol.owner
                 if (!ownerField.isStatic) {
-                    receiver = qualifiedAccess.findIrDispatchReceiver()
+                    // Qualified access with a selector: use the bound dispatch receiver.
+                    if (qualifiedAccess.explicitReceiver != null) {
+                        receiver = qualifiedAccess.findIrDispatchReceiver()
+                    } else {
+                        // Otherwise, find the dispatch receiver from the scope with a condition:
+                        //   the function in the scope has access to the field of interest.
+                        conversionScope.lastDispatchReceiverParameter { function ->
```

Comment:
IMHO this logic looks a bit hacky. I will think about alternative variants and write you.